src/coreclr/vm/methodtable.cpp

```diff
@@ -4219,28 +4220,92 @@ OBJECTREF MethodTable::GetManagedClassObject()
     {
         // Make sure that we have been restored
         CheckRestore();
+        AllocateRuntimeTypeObject(GetLoaderAllocator(), (RUNTIMETYPEHANDLE*)&GetWriteableData()->m_hExposedClassObject, TypeHandle(this));
+    }
+    RETURN(GetManagedClassObjectIfExists());
+}
 
-        REFLECTCLASSBASEREF  refClass = NULL;
-        GCPROTECT_BEGIN(refClass);
-        refClass = (REFLECTCLASSBASEREF) AllocateObject(g_pRuntimeTypeClass);
+//======================================================================================
+// Allocates a RuntimeType object with the given TypeHandle. If the LoaderAllocator
+// represents a not-unloadable context, it allocates the object on a frozen segment
+// so the direct reference will be stored to the pDest argument. In case of unloadable
+// context, an index to the pinned table will be saved.
+//======================================================================================
+void MethodTable::AllocateRuntimeTypeObject(LoaderAllocator* allocator, RUNTIMETYPEHANDLE* pDest, TypeHandle type)
+{
+    REFLECTCLASSBASEREF refClass = NULL;
 
-        LoaderAllocator *pLoaderAllocator = GetLoaderAllocator();
+    if (!allocator->CanUnload())
+    {
+        CrstHolder exposedClassLock(AppDomain::GetMethodTableExposedClassObjectLock());
 
-        ((ReflectClassBaseObject*)OBJECTREFToObject(refClass))->SetType(TypeHandle(this));
-        ((ReflectClassBaseObject*)OBJECTREFToObject(refClass))->SetKeepAlive(pLoaderAllocator->GetExposedObject());
+        if (*pDest == NULL)
+        {
+            // Allocate RuntimeType on a frozen segment
+            // Take a lock here since we don't want to allocate redundant objects which won't be collected
+
+            FrozenObjectHeapManager* foh = SystemDomain::GetFrozenObjectHeapManager();
+            Object* obj = foh->TryAllocateObject(g_pRuntimeTypeClass, g_pRuntimeTypeClass->GetBaseSize());
+            _ASSERTE(obj != NULL);
+            _ASSERTE((((SSIZE_T)obj) & 1) == 0);
+            refClass = (REFLECTCLASSBASEREF)ObjectToOBJECTREF(obj);
+            refClass->SetType(type);
+            *pDest = (RUNTIMETYPEHANDLE)obj;
+        }
+    }
+    else
+    {
+        GCPROTECT_BEGIN(refClass);
+        refClass = (REFLECTCLASSBASEREF)AllocateObject(g_pRuntimeTypeClass);
+        refClass->SetKeepAlive(allocator->GetExposedObject());
+        LOADERHANDLE exposedClassObjectHandle = allocator->AllocateHandle(refClass);
+        // Set the lowest bit, we use it GetPinnedManagedClassObjectIfExists as a fast detection of the unloadable context
+        exposedClassObjectHandle |= 1;
```

Comment:
If you set the bit here, you would need to clear the bit on the backout path. This problem will go away if you incorporate my other feedback.

Reply:
@jkotas could you please point me to that feedback, I am a bit lost.
the bit here is set for objects in unloadable contexts

Reply:
"We can keep the polarity of the low-bit to avoid this conditional jump."  from https://github.com/dotnet/runtime/pull/75573#discussion_r974757277

Reply:
From my understanding, I don't need to clear it - it's ignored here https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/loaderallocator.inl#L127 when it gets index out of it